cmd/update-aka-ms/update-aka-ms.go

```diff
@@ -0,0 +1,237 @@
+	cmd.Args = append(cmd.Args, flag.Args()...)
+	return executil.Run(cmd)
+}
+
+type akaMSLinkPair struct {
+	short, target string
+}
+
+func (p *akaMSLinkPair) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
+	s := xml.StartElement{
+		Name: xml.Name{Local: "AkaMSLink"},
+		Attr: []xml.Attr{
+			{Name: xml.Name{Local: "Include"}, Value: p.short},
+			{Name: xml.Name{Local: "TargetUrl"}, Value: p.target},
+		},
+	}
+	err := e.EncodeToken(s)
+	if err != nil {
+		return err
+	}
+	return e.EncodeToken(s.End())
+}
+
+func createLinkPairs(assets buildassets.BuildAssets) ([]akaMSLinkPair, error) {
+	v := assets.GetGoVersion()
+	// The partial versions that we want to link to a specific build.
+	// For example, 1.18-fips -> 1.18.2-1-fips.
+	partial := []string{
+		v.MajorMinor() + v.NoteWithPrefix(),
+		v.MajorMinorPatch() + v.NoteWithPrefix(),
+		// Also include the fully specified version. This lets people use a pretty link even if they
+		// do need to pin to a specific version.
+		v.Full(),
+	}
+
+	goSrcURLParts := strings.Split(assets.GoSrcURL, "/")
+	if len(goSrcURLParts) < 3 {
+		return nil, fmt.Errorf("unable to determine build number from %#q: not enough '/' segments to be an asset URL", assets.GoSrcURL)
+	}
+	buildNumber := goSrcURLParts[len(goSrcURLParts)-2]
+
+	urls := make([]string, 0, 3*(len(assets.Arches)+1))
+	for _, a := range assets.Arches {
+		urls = appendURLAndVerificationURLs(urls, a.URL)
+	}
+	urls = appendURLAndVerificationURLs(urls, assets.GoSrcURL)
+
+	pairs := make([]akaMSLinkPair, 0, len(urls)*len(partial))
+
+	for _, p := range partial {
+		for _, u := range urls {
+			urlParts := strings.Split(u, "/")
+			if len(urlParts) < 3 {
+				return nil, fmt.Errorf("unable to determine short link for %#q: not enough '/' segments to be an asset URL", u)
+			}
+			filename := urlParts[len(urlParts)-1]
+			// Make our aka.ms links more like official Go links: remove '.' between first parts.
+			if strings.HasPrefix(filename, "go.") {
+				filename = "go" + strings.TrimPrefix(filename, "go.")
+			}
+			f, err := makeFloatingFilename(filename, buildNumber, p)
+			if err != nil {
+				return nil, fmt.Errorf("unable to process URL %#q: %w", u, err)
+			}
+
+			pairs = append(pairs, akaMSLinkPair{
+				*latestShortLinkPrefix + f,
+				u,
+			})
+		}
+	}
+
+	return pairs, nil
+}
+
+func appendURLAndVerificationURLs(u []string, url string) []string {
+	u = append(u, url, url+".sha256")
+	if strings.HasSuffix(url, ".tar.gz") {
+		u = append(u, url+".sig")
+	}
+	return u
+}
+
+func makeFloatingFilename(filename, buildNumber, floatVersion string) (string, error) {
+	f := strings.ReplaceAll(filename, buildNumber, floatVersion)
+	if f == filename {
+		return "", fmt.Errorf("unable to find buildNumber %#q in filename %#q", buildNumber, filename)
+	}
+	return f, nil
+}
+
+func getPropsFileContent(pairs []akaMSLinkPair) (string, error) {
+	var b strings.Builder
```

Comment:
I'm having a really hard time figuring out how the ItemGroup field in your example works... even changing it to this, it still produces the same, good `<Project><ItemGroup><AkaMSLink...` result: https://play.golang.com/p/5DdPc2t44NK

```go
	ItemGroup []akaMSLinkPair `xml:"IgnoredStuff >GoesHere"`
```

After reading https://pkg.go.dev/encoding/xml#Marshal a bit more, this is what I ended up with--was able to get rid of the unused `xml.Name` field in `akaMSLinkPair` (because I don't care about being able to marshal individual `akaMSLinkPair` outside the context of `akaMSPropsFile`) and not repeat any field names:

```go
type akaMSLinkPair struct {
	Short  string `xml:"Include,attr"`
	Target string `xml:"TargetUrl,attr"`
}

type akaMSPropsFile struct {
	XMLName   xml.Name        `xml:"Project"`
	ItemGroup []akaMSLinkPair `xml:">AkaMSLink"`
}
```

The docs only have one brief sentence about `>`, and it doesn't mention spaces as far as I can tell, so I'm still not sure what's up with `IgnoredStuff >GoesHere` being effectively ignored and using the exported field's name.

Reply:
> I'm having a really hard time figuring out how the ItemGroup field in your example works... even changing it to this, it still produces the same, good <Project><ItemGroup><AkaMSLink... result: https://play.golang.com/p/5DdPc2t44NK

> > ItemGroup []akaMSLinkPair `xml:"IgnoredStuff >GoesHere"`

Hmm, my take is that marshaling this struct should error our, but it doesn't. The relevant code is this: https://github.com/golang/go/blob/8ab42a945aef7c09189a60263c4450a65bc00a47/src/encoding/xml/typeinfo.go#L118

It is treating `IgnoredStuff` as the namespace and ` >GoesHere` as the tag. Although there is no documentation on how this should behave, is we go down a little bit in the code, there is a check which should catch namespaces without names:

https://github.com/golang/go/blob/8ab42a945aef7c09189a60263c4450a65bc00a47/src/encoding/xml/typeinfo.go#L174